MainWindow.cpp

```diff
@@ -469,8 +469,35 @@ void MainWindow::on_actionOpen_triggered() {
 }
 
 void MainWindow::on_actionSave_triggered() {
-  const QString &filter = "EGM (*.egm)";
-  const QString &fileName = QFileDialog::getSaveFileName(this, tr("Save Project"), "", filter);
+  // map useful to quickly fetch extension from the selected filter in QDialog (because Qt doesn't give a better way)
+  std::unordered_map<QString, QString> extensionMap;
+  extensionMap["EGM (*.egm)"] = ".egm";
+
+  // lambda to create a QString of all filters separated by ";;" except last. The generated QString is required
+  // as an argument in getSaveFileName
+  auto getFilters = [&extensionMap]() -> QString {
+    assert(static_cast<int>(extensionMap.size()) > 0);
+
+    auto itr = extensionMap.begin();
+    QString f = itr->first;
+    ++itr;
+
+    for(; itr != extensionMap.end(); ++itr){
+      f += ";;" + itr->first;
+    }
+    return f;
+  };
+
+  QString selectedFilter;
+  QString fileName = QFileDialog::getSaveFileName(this, tr("Save Project"), "", getFilters(), &selectedFilter);
+
+  // to check if fileName contains extension
+  const std::filesystem::path &filePath = fileName.toStdString();
+  const std::string &ext = ToLower(filePath.extension().u8string());
+  if (ext.empty()) {

```

Comment:
All this conversion just to check the extension? And why lowercasing for empty check? If we stay with Qt API, we just do:
```c++
if (QFileInfo(fileName).suffix().isEmpty()) {
```
However, this will still fail to add an extension to any files with a dot in their name, or ending with a dot. I think we just want to make sure to add the extension if needed. Maybe something like:
```c++
if (!fileName.endsWith(extensionMap[selectedFilter], Qt::CaseInsensitve)) {
    while (fileName.endsWith(QLatin1Char('.'))
        fileName.chop(1);
    fileName.append(extensionMap[selectedFilter]);
}
```

Reply:
You're right ToLower is not useful here, I copy-pasted it from other part and overlooked the usage. I've made the changes!